### ISSUE TITLE
.NET Core support and CoreCLR embedding - cross-platform API

### ISSUE THREAD

**Issue by denfromufa** [pythonnet/pythonnet#96]
Looks like with CoreCLR the unmanaged exports "hack" for embedding CLR can be extended by cross-platform hosting API:

https://github.com/dotnet/coreclr/issues/1256

This would allow simple C/C++ code to embed CLR

Current state of .NET Core support:

https://github.com/pythonnet/pythonnet/issues/96#issuecomment-430465993

**Comment by denfromufa** [comment]
Note that for .NET 4+ still have to use unmanaged exports.

**Comment by denfromufa** [comment]
apparently @lstratman has extensive experience porting Mono to CoreCLR embedding.

**Comment by lstratman** [comment]
Thanks @denfromufa :).

It's funny, I actually had a fork of this repo from a while back with the goal of looking into this.  But, since my company has only a single customer that might want to integrate our .NET server application into their Python application on Linux and, even then, only at some point in the future, it's down towards the bottom of my priority list for development unfortunately.  However, I'm happy to offer some technical guidance.

I implemented basically this same functionality for Node.js for the [Edge.js project](https://github.com/tjanczuk/edge) and a lot of the code that I wrote there can probably be re-used or easily re-purposed for pythonnet.  https://github.com/tjanczuk/edge/blob/master/src/CoreCLREmbedding/coreclrembedding.cpp#L192 contains the code that spins up the .NET Core CLR in any process.  Towards the end of the method (https://github.com/tjanczuk/edge/blob/master/src/CoreCLREmbedding/coreclrembedding.cpp#L673), we get pointers to a bunch of .NET methods that serve as the managed end of the Node.js->.NET interop process.  These function pointers are then invoked anytime that we need to call from Node.js to .NET.  Implementations of those .NET methods are in https://github.com/tjanczuk/edge/blob/master/src/double/Edge.js/dotnetcore/coreclrembedding.cs and are included in the .NET Core application that's run from within Edge.js through a NuGet package that needs to be referenced (see https://github.com/tjanczuk/edge#how-to-specify-additional-clr-assembly-references-in-c-code).

This works well for Edge.js because the interaction between Node.js and .NET is all encapsulated in functions in one language or the other.  So to call between the two domains, you're always invoking functions, never creating variables, setting properties, etc.  However for pythonnet, we're obviously more "integrated" into the Python syntax, so we can do things like create variables, set properties, etc.  You can write more methods in the .NET end of the interop process to facilitate this, or try to stay in unmanaged code as much as possible by calling the .NET Core CLR native methods to do things like create variables and set properties.  That would be the more performant, but potentially more difficult, route.  This will have to be done through the ICLRRuntimeHost instance that you get back from calling coreclr_initialize().  It should theoretically be possible, but is left as an exercise to the interested reader.

Anyway, that concludes my brain dump, I hope that it helps.  Feel free to post questions and I'll do my best to answer them.  At some point, I may have some time for actual development as well, but can't offer any promises.

**Comment by denfromufa** [comment]
@lstratman thank you for very detailed response, there is a question from MS if CoreCLR C-API is good enough, would you mind replying there? MS is considering standardizing C-API for .NET ecosystem.

https://blogs.msdn.microsoft.com/dotnet/2016/09/26/introducing-net-standard/#comment-138395


Are the four methods (Initialize, Shutdown, Create_Delegate, Execute_Assembly), mentioned enough for embedding CoreCLR in pythonnet?

**Comment by lstratman** [comment]
I've replied to that thread that those methods are sufficient.  Looking further into the existing pythonnet code, I realized that it does all of the things that I mentioned earlier (create variables, set properties, index arrays, etc.) in managed code anyway, so once the "spinning up CoreCLR" code is integrated from Edge.js, pythonnet should just be able to re-use a lot of its existing codebase.

**Comment by denfromufa** [comment]
Another thing holding of CoreCLR port is Reflection.Emit used in pythonnet in one place:

https://github.com/dotnet/corefx/issues/4491#issuecomment-189756092

But we could possibly just do conditional compilation for this feature.

**Comment by denfromufa** [comment]
otherwise try ikvm.reflection.emit or mono.cecil

**Comment by var1ap** [comment]
Hi, I started work toward Python.Net Core edition in this fork https://github.com/var1ap/pythonnet , stay tuned.

**Comment by denfromufa** [comment]
@var1ap this is great news, do you agree to transition from Zope to MIT license, if your port is successful?

**Comment by var1ap** [comment]
@denfromufa of course agree.

**Comment by var1ap** [comment]
greate news!
https://github.com/var1ap/pythonnet/commit/2b508fd3fda1f67db6785b03a409b9edd0e67686
i managet to get pythonnet work under dotnet core 1.1(Win x64)

**Comment by var1ap** [comment]
Also pythonnet successfully ported to dotnet core 1.1(Linux x64 Ubuntu 16.04) (in the same fork).
This is just Proof of concept. So somebody should port this project to CoreCLR with good code quality required for production projects.

**Comment by dmitriyse** [comment]
Hi guys! Pythonnet really works under dotnet core >= 1.0(NetStandard1.5). Thank you very much var1ap for your research. I continued to work on porting pythonnet to dotnet core in this repo https://github.com/dmitriyse/pythonnet/tree/coreclr. Also I plan to improve my binary distribution and avoid any dll load magic. Finally we should get good nuget based distribution.

**Comment by denfromufa** [comment]
@var1ap @dmitriyse can you prepare a pull request with compilation flags for coreclr? much like .NET/Mono flags that we have currently.

**Comment by dmitriyse** [comment]
I can spend time to reapply coreclr code adoption in the next week.
Please review talk in those posts https://github.com/dmitriyse/pythonnet/commit/38adb2e8701f78a203ca5f7ddc60582c382e2b6b#commitcomment-20774083

Please consider with vmuriart, which fork/branch is most suitable for this work, where I can fork.

**Comment by vmuriart** [comment]
@dmitriyse on top of `master`, but at the rate we are committing changes at the moment it wouldn't make sense to do all the work if it won't get merged in immediately. 

@denfromufa, after reviewing @dmitriyse and @var1ap, their changes look pretty straight forward to upgrade to `coreclr`. My hesitation is that it seems that it would require us to upgrade from `4.0` to `4.5` framework which may drop users (I still use 4.0). We could throw in a bunch of `#if` directives, but it would make the code very messy. (side note, We actually got rid of most `mono_linux` and `mono_osx` flags from the code thanks to @dmitriyse)

Another issue is that `coreclr` is still undergoing changes wrt their build system. They've gone from `project.json` back to an upgraded `msbuild` with new project file formats. `mono` wouldn't compile against either of these options; ie. we would need to have two sets of `csproj`/`sln`/`packages.config` files in order to ensure `mono` can still be used for compiling (unless we drop `mono`?)

Currently what I'm doing is slowly porting over some of the changes @dmitriyse and @var1ap did to minimize the changes for when we migrate to `coreclr`. I'm hoping to find a clean way to maintain `clr 4.0` and add `coreclr`, but at the moment I'm not finding it and it might need to be part of `pythonnet 3.0`.

**Comment by denfromufa** [comment]
I agree with your comments @vmuriart

.NET 4.5 is a bummer, I'm also mostly on .NET 4.0 due to old Windows Server
editions.

I'm also against deprecating Mono, at least until CoreCLR works better.

**Comment by dmitriyse** [comment]
Hi! I restarted work on porting PythonNet to CoreCLR. Now it can be performed as additional files.
So current build and new build (based on .csproj 2017) can coexist side-by-side.
https://github.com/dmitriyse/pythonnet/tree/coreclr

**Comment by denfromufa** [comment]
@dmitriyse this looks much cleaner! Can you open a pull request for review cycle?

![image](https://cloud.githubusercontent.com/assets/7870949/24414929/6575aaf6-13a5-11e7-8228-8f3527a8cc9d.png)

**Comment by dmitriyse** [comment]
Ok, I will create pool request right after some useful point. I fixed 102 from 120 migration issues to get pythonnet working with at least one build configuration.

**Comment by dmitriyse** [comment]
Half of the problem solved - python can be embedded in to CoreCLR 2.0 (https://github.com/pythonnet/pythonnet/pull/519)

System.Reflection.Emit works good under CoreCLR 2.0 except for CDecl calling convention. CDecl currently unsupported in the CoreCLR 2.0

As a workaround I changed  XDecref/XIncref to Py_DecRef/, Py_IncRef.
And used UnamanagedFunctionPoinerAttribute with Marshal.GetDelegateForFunctionPointer, with works good, and does not looks too slow.

The rest part is to inject CoreCLR into python through CoreCLR C Api.

**Comment by denfromufa** [comment]
@dmitriyse for cdecl did you mean this issue? https://github.com/dotnet/corefx/issues/9800

**Comment by dmitriyse** [comment]
Yes, exactly this flaw.

**Comment by denfromufa** [comment]
@dmitriyse for rest part to inject CoreCLR into Python we can avoid CoreCLR C API, if the DLLExport becomes stable with .NET Core, @3F has to use custom CoreCLR at the moment. DLLExport is the next generation of UnmanagedExports, that we are currently using in .NET Framework:

https://github.com/3F/DllExport

**Comment by denfromufa** [comment]
@dmitriyse with your valid finalizer from #532 I'm getting this unexpected crash on .NET Core 2.0, which does not appear on .NET Framework:

```
C:\Users\denis.akhiyarov\Downloads\pythonnet-valid-finalizer\build\lib.win32-2.7\netcoreapp2.0>dotnet nPython.dll
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:42:59) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import clr
>>> import System
>>> System.Double(1.0)

Unhandled Exception: System.InvalidCastException: Unable to cast object of type 'TernaryFunc' to type 'IntPtr_3_Delegate'.
   at Python.Runtime.NativeCall.Call_3(IntPtr fp, IntPtr a1, IntPtr a2, IntPtr a3)
   at Python.Runtime.MetaType.tp_call(IntPtr tp, IntPtr args, IntPtr kw)
   at Python.Runtime.Runtime.Py_Main(Int32 argc, String[] argv)
   at Python.Runtime.PythonConsole.Main(String[] args)

C:\Users\denis.akhiyarov\Downloads\pythonnet-valid-finalizer\build\lib.win32-2.7\netcoreapp2.0>cd..

C:\Users\denis.akhiyarov\Downloads\pythonnet-valid-finalizer\build\lib.win32-2.7>nPython
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:42:59) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import clr
>>> import System
>>> System.Double(1.0)
<System.Double object at 0x052510B0>
```

**Comment by dmitriyse** [comment]
Thank you for the case. I will try to reproduce and debug.

**Comment by dmitriyse** [comment]
I found the problem and solution. It's a DotNet core bug. I will create PR soon.

**Comment by denfromufa** [comment]
some progress on .NET Core version of DllExport:

https://github.com/3F/DllExport/issues/17#issuecomment-333971688

**Comment by denfromufa** [comment]
@dmitriyse another approach we can take (if DLLExport has technical limitations, and C-API is too hard), is using Embeddinator project from Mono, which supposedly also works under .NET Standard / .NET Core:

https://github.com/mono/Embeddinator-4000/issues/515

**Comment by dmitriyse** [comment]
I am afraid, that Embeddinator-4000 uses mono/.net framework under the hood. We need to inspect example with .Net Standard 1.4 library to ensure that DotNotCore 2.0 host can be used.

**Comment by denfromufa** [comment]
@dmitriyse here is why I think this is possible:

https://github.com/mono/CppSharp/issues/613

**Comment by Cronan** [comment]
Any updates on this issue? As per a discussion with @denfromufa, it seems this might hold the key: https://developer.xamarin.com/guides/cross-platform/dotnet-embedding/getting-started/c/

**Comment by denfromufa** [comment]
@Cronan very disappointing, but I just found out that .NET Core is not implemented in embeddinator. Hence we need to use this cross-platform C/C++ hosting API for .NET Core like it is currently done for Mono:

https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting

**Comment by Cronan** [comment]
Is this similar to what they did in edge.js?

**Comment by Cronan** [comment]
@denfromufa @dmitriyse I've got a stand-alone test project that successfully calls a method in a dotnet core 2.0 app, using the cross-platform C/C++ hosting API for .Net Core 2.0. The code uses dlopen() and dlsym() to initialize the core CLR runtime, and call a function using a delegate.

It's not really a branch, more of a stand-alone piece of work for now, but I hope it won't be too much work to integrate into the library.

Where should I put it so you can take a look?

**Comment by denfromufa** [comment]
Just attach it here in this issue as a zip archive. This is a good news!

**Comment by Cronan** [comment]
# Build
```
dotnet build ./dotnet
python setup.py build_ext --inplace
```

# Run
```
python
>>> import clr
>>> clr.sum(4,5)
9.0
```
[coreclr_host.zip](https://github.com/pythonnet/pythonnet/files/1681015/coreclr_host.zip)

**Comment by djoyce82** [comment]
@denfromufa @Cronan I have been working on a fork of this here:

https://github.com/djoyce82/pythonnet

It's very similar to what @Cronan has put together (similar reference material :-) ) but I have ported it to C as this is what the current mono implementation appears to be targeting.

I've tested this on Linux (CentOS 7) and the code works for my use case.

Would you like me to create a pull request for these changes?

**Comment by denfromufa** [comment]
@djoyce82 this is great! definitely, please submit a pull request.

**Comment by Cronan** [comment]
@djoyce82 That's great work, does it build and run from Linux? What commands did you use to build it and run the tests?

**Comment by Cronan** [comment]
@djoyce82 

```
python setup.py bdist_wheel --xplat
```

```
Microsoft (R) Build Engine version 15.3.409.57025 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Python.Runtime.15 -> /users/isys/icronyn/git/pythonnet_clr/src/runtime/bin/netstandard2.0/Python.Runtime.dll
  Console.15 -> /users/isys/icronyn/git/pythonnet_clr/src/console/bin/netcoreapp2.0/nPython.dll
  Console.15 -> /users/isys/icronyn/git/pythonnet_clr/build/lib.linux-x86_64-2.7/netcoreapp2.0/
  Python.EmbeddingTest.15 -> /users/isys/icronyn/git/pythonnet_clr/src/embed_tests/bin/netcoreapp2.0/Python.EmbeddingTest.dll
  Python.EmbeddingTest.15 -> /users/isys/icronyn/git/pythonnet_clr/src/embed_tests/bin/netcoreapp2.0_publish/
building 'clr' extension
creating build/temp.linux-x86_64-2.7
creating build/temp.linux-x86_64-2.7/src
creating build/temp.linux-x86_64-2.7/src/coreclr
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/man/releases/python-medusa/27-1/include/python2.7 -c src/coreclr/pynetinit.c -o build/temp.linux-x86_64-2.7/src/coreclr/pynetinit.o
src/coreclr/pynetinit.c: In function \u2018init\u2019:
src/coreclr/pynetinit.c:112:5: error: \u2018for\u2019 loop initial declarations are only allowed in C99 mode
     for (int ii = 0; ii < PyList_Size(syspath); ++ii)

```

**Comment by denfromufa** [comment]
@Cronan pass `'-std=c99'` to compile args:

https://stackoverflow.com/a/10867041/2230844

**Comment by Cronan** [comment]
Thanks @denfromufa, I fixed it by declaring the variable outside the loop - there was a single other case of this:

```
src/coreclr/coreutils.c: In function \u2018AddFilesFromDirectoryToTpaList\u2019:
src/coreclr/coreutils.c:236:5: error: \u2018for\u2019 loop initial declarations are only allowed in C99 mode
     for (size_t extIndex = 0; extIndex < sizeof(tpaExtensions) / sizeof(tpaExtensions[0]); extIndex++)
     ^
src/coreclr/coreutils.c:236:5: note: use option -std=c99 or -std=gnu99 to compile your code
error: command 'gcc' failed with exit status 1
```

**Comment by Cronan** [comment]
@denfromufa @djoyce82 Lots of the tests failed, with this kind of thing:

```
Unhandled Exception: System.MissingMethodException: Constructor on type 'System.Reflection.Emit.TypeBuilder' not found.
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(Type type, Object[] args)
   at Python.Runtime.DelegateManager.GetDelegate(Type dtype, IntPtr callable)
   at Python.Runtime.EventObject.AddEventHandler(IntPtr target, IntPtr handler)
   at Python.Runtime.EventBinding.nb_inplace_add(IntPtr ob, IntPtr arg)
Aborted
```

**Comment by djoyce82** [comment]
@Cronan I believe the test failures are the same issue as detailed in #590 and #571, looks like the implementation of System.Reflection.Emit.TypeBuilder has changed in .NET Core?

PR is #612

**Comment by weixu02** [comment]
If I want to use a .net core compiled lib in python and only care about using on windows, is that possible with the current python.net version? Thanks!

**Comment by denfromufa** [comment]
@weixu02 this is possible via npython.exe if you compile the full pythonnet solution

**Comment by MatthiasKauer** [comment]
Hi, I'm trying to test calling from dotnet core into CPython.
If I understand @denfromufa correctly, that should be working.
But how do I set it up? `pip install pythonnet` fails with the following message:
    
> Could not load file or assembly 'System.ComponentModel.Composition, Version=4.0.0.0,Culture=neutral, PublicKeyToken=b77a5c561934e089' or one of its dependencies.

Installing the source version with `python setup.py build_ext --inplace` or just executing `mono tools/nuget/nuget.exe` leads to the same error. My interpretation attempt: `mono` on my Linuxmint 18.2 system could be too old.

I also tried installing via nuget but got (for framework netcoreapp2.0)
> error: Package 'pythonnet' is incompatible with 'all' frameworks in project '/home/matt/dvl/experiments/pythonnet-core/pythonnet-core.csproj'.

Questions:
- Is upgrading `mono` through their apt repository the way to go?
- Does the status table from @denfromufa maybe reflect the current master but not what is currently released?
- If I want to write dotnet around pythonnet do I still install it via `pip` not via nuget?

**Comment by MrM40** [comment]
Sorry for my ignorance. I'm trying to get embedding to work. I've installed pythonnet from nuget (2.3.0-py27-dotnet). But this **Python.Runtime.dll** doesn't seem to be .Net Standard or -Core. Tried to compile the source code  **\pythonnet-master\src\runtime\Python.Runtime.csproj** to Standard or Core but it seem to be a .Net 4.x. Also tried the Python.Runtime.dll in the Wheel package from [here.](https://pypi.org/project/pythonnet/#files), same problem.
Can someone point me in a direction on how to get my hands on a .Net Standard or -Core version of Python.Runtime.dll?
I read a lot of places that the master branch support .Core....but I really don't know what to do with this information beside what I already tried?

**Comment by denfromufa** [comment]
@MrM40 .net core builds are in this solution:

https://github.com/pythonnet/pythonnet/blob/master/pythonnet.15.sln

But I recommend building from python side by passing the `--xplat` flag to setup.py:

https://github.com/pythonnet/pythonnet/blob/master/appveyor.yml#L60

You can also grab the builds/wheels from appveyor artifacts.

It is not possible to keep all 3 major runtimes (.NET Framework, .NET Core, and Mono) in the same build.

**Comment by MrM40** [comment]
Still running against a wall. I only see release/debug to Mono and Win in those projects?
And I assume I need to compile the "Console.15" project, right
By the way, why not target .NET Standard in these projects?

**Comment by denfromufa** [comment]
@MrM40 see this: https://github.com/pythonnet/pythonnet/blob/master/src/runtime/Python.Runtime.15.csproj#L4

**Comment by MrM40** [comment]
Thanks :-)
I'm running on Linux(Debian)  and now I get this errir:

`Start
Unhandled Exception: System.DllNotFoundException: Unable to load DLL 'python27': The specified module or one of its dependencies could not be found.
 (Exception from HRESULT: 0x8007007E)
   at Python.Runtime.Runtime.Py_IsInitialized()
   at Python.Runtime.Runtime.Initialize(Boolean initSigs) in D:\Development\~.Net libraries (3.part)\phytonnet\2018-12-09\pythonnet-master\src\runtime\runtime.cs:line 274
   at Python.Runtime.PythonEngine.Initialize(IEnumerable1 args, Boolean setSysArgv, Boolean initSigs) in \pythonnet-master\src\runtime\pythonengine.cs:line 169
   at Python.Runtime.Py.GIL() in \pythonnet-master\src\runtime\pythonengine.cs:line 619
   at pythonnet_w.Program.Main(String[] args) in \pythonnet_test\Program.cs:line 9
Aborted
`

I made two links: **python2.7** and **python27** in the same folder as **Python.Runtime.dll** pointing to  "/usr/bin/python2.7".

Not I got this error:

`Start
Unhandled Exception: System.EntryPointNotFoundException: Unable to find an entry point named 'PyUnicodeUCS2_FromUnicode' in DLL 'python27'.
   at Python.Runtime.Runtime.PyUnicode_FromUnicode(String s, IntPtr size)
   at Python.Runtime.Runtime.Initialize(Boolean initSigs) in \pythonnet-master\src\runtime\runtime.cs:line 337
   at Python.Runtime.PythonEngine.Initialize(IEnumerable 1 args, Boolean setSysArgv, Boolean initSigs) in pythonnet-master\src\runtime\pythonengine.cs:line 169
   at Python.Runtime.Py.GIL() in \pythonnet-master\src\runtime\pythonengine.cs:line 619
   at pythonnet_w.Program.Main(String[] args) in \pythonnet_test\Program.cs:line 9
Aborted`

**Comment by denfromufa** [comment]
@MrM40 please don't sidetrack this .NET Core (CoreCLR) status with other issues. In a separate issue, document how you are building and how you are using pythonnet to make a reproducible case. Also search the issue tracker for duplicate issues before posting.

**Comment by alumis** [comment]
I built https://github.com/pythonnet/pythonnet/blob/master/pythonnet.15.sln with the following compilation symbols:

`XPLAT;;PYTHON3;PYTHON37;UCS4;NETSTANDARD;`

Running this in a .NET Core application on Windows 8.1 causes the program to fatally crash:

`The program '[11180] dotnet.exe' has exited with code -1073740791 (0xc0000409).`

Without any more information. I'm sure more information can be obtained, but I'm unsure how.

Kind regards, Emil

**Comment by dmitriyse** [comment]
Please try UCS2

**Comment by alumis** [comment]
@dmitriyse Thanks! Changed to UCS2, now getting:

`The program '[12304] dotnet.exe' has exited with code 255 (0xff).`

**Comment by dmitriyse** [comment]
Unfortunately, I haven't tested this solution widely. Win 8.1 setup hasn't been tested. So you need to dive deeply with VS debugging.

**Comment by alumis** [comment]
@dmitriyse Hah! I got it working :) I had this line hanging around:

> PythonEngine.PythonPath = @"C:\Users\E\AppData\Local\Programs\Python\Python37\";

Removing it fixed the issue.

**Comment by filmor** [comment]
Closed in favour of #984.
